lib/datadog/core/utils/network.rb

```diff
@@ -0,0 +1,160 @@
+
+          def global_ip?(parsed_ip)
+            parsed_ip && !private?(parsed_ip) && !loopback?(parsed_ip) && !link_local?(parsed_ip)
+          end
+
+          # vendor code from https://github.com/ruby/ipaddr/blob/master/lib/ipaddr.rb
+          # Ruby version below 2.5 does not have the IpAddr#private? method
+          # We have to vendor the code because ruby versions below 2.5 did not extract ipaddr as a gem
+          # So we can not specify a specific version for ipaddr for ruby versions: 2.1, 2.2, 2.3, 2.4
+          # TODO: remove once we drop support for ruby 2.1, 2.2, 2.3, 2.4
+          def private?(ip)
+            addr = ip.instance_variable_get(:@addr)
+
+            case ip.family
+            when Socket::AF_INET
+              addr & 0xff000000 == 0x0a000000 ||    # 10.0.0.0/8
+                addr & 0xfff00000 == 0xac100000 ||  # 172.16.0.0/12
+                addr & 0xffff0000 == 0xc0a80000     # 192.168.0.0/16
+            when Socket::AF_INET6
+              addr & 0xfe00_0000_0000_0000_0000_0000_0000_0000 == 0xfc00_0000_0000_0000_0000_0000_0000_0000
+            else
+              raise IPAddr::AddressFamilyError, 'unsupported address family'
+            end
+          end
+
+          # vendor code from https://github.com/ruby/ipaddr/blob/master/lib/ipaddr.rb
+          # Ruby version below 2.5 does not have the IpAddr#link_local? method
+          # We have to vendor the code because ruby versions below 2.5 did not extract ipaddr as a gem
+          # So we can not specify a specific version for ipaddr for ruby versions: 2.1, 2.2, 2.3, 2.4
+          # TODO: remove once we drop support for ruby 2.1, 2.2, 2.3, 2.4
+          def link_local?(ip)
+            addr = ip.instance_variable_get(:@addr)
+
+            case ip.family
+            when Socket::AF_INET
+              addr & 0xffff0000 == 0xa9fe0000 # 169.254.0.0/16
+            when Socket::AF_INET6
+              addr & 0xffc0_0000_0000_0000_0000_0000_0000_0000 == 0xfe80_0000_0000_0000_0000_0000_0000_0000
+            else
+              raise IPAddr::AddressFamilyError, 'unsupported address family'
+            end
+          end
+
+          # vendor code from https://github.com/ruby/ipaddr/blob/master/lib/ipaddr.rb
+          # Ruby version below 2.5 does not have the IpAddr#loopback? method
+          # We have to vendor the code because ruby versions below 2.5 did not extract ipaddr as a gem
+          # So we can not specify a specific version for ipaddr for ruby versions: 2.1, 2.2, 2.3, 2.4
+          # TODO: remove once we drop support for ruby 2.1, 2.2, 2.3, 2.4
+          def loopback?(ip)
+            addr = ip.instance_variable_get(:@addr)
+
+            case ip.family
+            when Socket::AF_INET
+              addr & 0xff000000 == 0x7f000000
+            when Socket::AF_INET6
+              addr == 1
+            else
+              raise IPAddr::AddressFamilyError, 'unsupported address family'
+            end
+          end
```

Comment:
Since this is third-party code imported from an external gem, I suggest isolating it on its own separate file (that we can then remove when we decide to drop support for the older Rubies).

Furthermore, remember to:
* Update `LICENSE-3rdparty.csv` to mention this code (required by Datadog)
* Add https://github.com/ruby/ipaddr/blob/master/LICENSE.txt as a header to the file containing this code (required by the license -- "Redistribution [...] in source [...] with or without modification [...] must retain the above copyright notice, this list of conditions and the following disclaimer [...]")

Reply:
Moreover, to @ivoanjo 's point (which is a very good one), it should be moved entirely to a "vendor" folder. We want to create a physical separation between our code and the 3rd party code we don't maintain. This helps prevent coupling, misidentification (as Datadog code) and makes it easier to externally label which code is ours and which isn't (by simply stating "all 3rd party code is in the 'vendor' path".)